Fix blank lines between list items and nested sub-lists being removed in Markdown/MDX

byplayer opened the pull request:
## Description

This is a bug fix of issue #17746.

Preserve blank lines in nested lists

Fixed an issue where prettier was removing blank lines between list items
and nested sublists in Markdown, which could change the semantic meaning
of the content. The fix detects original blank lines in the source and
preserves them when they occur before nested lists within list items.

This addresses cases where users intentionally use blank lines to create
visual separation or semantic grouping in their markdown lists.

## Checklist

<!-- Please ensure you’ve done all of these things (if applicable). -->
<!-- You can replace the `[ ]` with `[x]` to mark each task as done. -->

- [x] I’ve added tests to confirm my change works.
- [x] (If changing the API or CLI) I’ve documented the changes I’ve made (in the `docs/` directory).
- [x] (If the change is user-facing) I’ve added my changes to `changelog_unreleased/*/XXXX.md` file following `changelog_unreleased/TEMPLATE.md`.
- [x] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/main/CONTRIBUTING.md).

byplayer committed:
🐛 fix(markdown): preserve blank lines in nested lists

Fixed an issue where prettier was removing blank lines between list items
and nested sublists in Markdown, which could change the semantic meaning
of the content. The fix detects original blank lines in the source and
preserves them when they occur before nested lists within list items.

This addresses cases where users intentionally use blank lines to create
visual separation or semantic grouping in their markdown lists.

Fixes #17746

byplayer committed:
📝 Add changelog for markdown blank line fix (#17746)

Add changelog entry for the fix that preserves blank lines in
nested markdown lists, addressing issue #17746.

This changelog documents the bug fix that ensures markdown formatting
maintains proper spacing in nested list structures, improving output
consistency with markdown standards.

byplayer committed:
✅ update snapshots for markdown blank line fix

Why is this change needed?
The test snapshots need to be updated to reflect the fix for issue #17746, which preserves blank lines in nested markdown lists. The snapshots now correctly show blank lines between numbered list items and their nested sub-items.

How does it address the issue?
Updates the Jest snapshot file to match the corrected formatting behavior where blank lines are preserved in nested list structures, ensuring tests pass with the new formatting rules.

What effects does the patch have?
Test snapshots now accurately represent the expected markdown formatting output, ensuring that the markdown parser correctly maintains blank lines in nested list structures as intended by the fix.

byplayer committed:
🔀 Merge branch 'main' into fix-17746

andersk commented:
The merge conflict with #18321 needs to be resolved.

tats-u commented:
@byplayer You have never got a review yet and PRs are squashed when merged, so you can either `merge main` or `rebase main`. `merge` is easier because of less conflicts.

byplayer committed:
Merge remote-tracking branch 'origin/main' into fix-17746

# Conflicts:
#	src/language-markdown/print/mdast.js
#	tests/format/markdown/commonmark-test-suite/__snapshots__/format.test.js.snap

byplayer committed:
🐛 fix(markdown): port blank line fix to children.js

After merging main's refactoring (which split printer-markdown.js into
separate files), the blank line preservation logic for issue #17746 was
re-applied to its new location in children.js.

# Why is this change needed?
Main's refactoring moved shouldPrePrintDoubleHardline to children.js.
The fix that preserves blank lines before nested lists within list items
was not carried over in the refactoring, causing the behavior to regress.

# How does it address the issue?
Added a check in shouldPrePrintDoubleHardline (non-MDX path) to return
true when there is a blank line in the original source between a paragraph
and a nested list within a listItem.

byplayer committed:
🐛 fix(markdown): don't add blank line after tight code+list

# Why is this change needed?
When a listItem contained an indented code block immediately followed by
a nested list (no blank line in original), prettier was incorrectly
inserting a blank line before the nested list, converting a tight list
to a loose one and changing the rendered output.

# How does it address the issue?
Added a blank line presence check (previous.position.end.line + 1 <
node.position.start.line) to the "code before list" branch in
shouldPrePrintDoubleHardline so it only preserves blank lines that
actually existed in the original source.

byplayer committed:
🐛 fix(markdown): preserve blank lines before nested list in MDX

The same loose-list bug (#17746) existed in the MDX parser path.
The `shouldPrePrintDoubleHardline` function had a `paragraph → list`
blank-line check added to the non-MDX branch but the MDX branch was
missed.

Added the equivalent check to the MDX branch, a new MDX test fixture
(issue-17746.mdx), and updated the changelog to mention MDX.

byplayer committed:
🐛 fix(markdown): guard MDX code→list with line-gap check

The MDX branch of `shouldPrePrintDoubleHardline` was inserting a blank
line before a nested list after a code node even when no blank line
existed in the source. Added the same `position.end.line + 1 <
position.start.line` guard already present in the non-MDX branch.

Added `issue-17746-code-before-list.mdx` as a regression test to
confirm tight code→list stays tight in the MDX parser path.

byplayer committed:
📝 docs(markdown): improve changelog for #17746

Make the changelog title more concrete and self-contained.
Add a description and a code example showing the before/after
behavior of the blank-line preservation fix for loose lists
with nested sublists in Markdown/MDX.

byplayer committed:
📝 docs(markdown): fix spellcheck error in changelog for #17746

Replace "sublists" with "sub-lists" in the changelog entry to pass
the cSpell spellcheck lint.

tats-u commented:
Could you change the title of this PR to that written in your changelog?

byplayer commented:
> Could you change the title of this PR to that written in your changelog?

I updated it.

tats-u commented on src/language-markdown/print/children.js:
```diff
@@ -71,7 +71,18 @@ function shouldPrePrintDoubleHardline(path, options) {
       (node.type === "list" &&
         parent.type === "listItem" &&
-        previous.type === "code")
+        previous.type === "code" &&
+        previous.position.end.line + 1 < node.position.start.line)
+    ) {
+      return true;
+    }
+
+    // Preserve blank line before nested list within listItem (issue #17746)
+    if (
+      node.type === "list" &&
+      parent.type === "listItem" &&
+      previous.type === "paragraph" &&
+      previous.position.end.line + 1 < node.position.start.line
```
```suggestion
        (previous.type === "code" ||
          // Preserve blank line before nested list within listItem (issue #17746)
          previous.type === "paragraph") &&
        previous.position.end.line + 1 < node.position.start.line)
```

tats-u commented on src/language-markdown/print/children.js:
```diff
@@ -87,7 +98,18 @@ function shouldPrePrintDoubleHardline(path, options) {
       (node.type === "list" &&
         parent.type === "listItem" &&
-        previous.type === "code")
+        previous.type === "code" &&
+        previous.position.end.line + 1 < node.position.start.line)
+    ) {
+      return true;
+    }
+
+    // Preserve blank line before nested list within listItem (issue #17746)
+    if (
+      node.type === "list" &&
+      parent.type === "listItem" &&
+      previous.type === "paragraph" &&
+      previous.position.end.line + 1 < node.position.start.line
```
```suggestion
        (previous.type === "code" ||
           // Preserve blank line before nested list within listItem (issue #17746)
           previous.type === "paragraph") &&
        previous.position.end.line + 1 < node.position.start.line)
```

tats-u reviewed (commented):
Two `if` statements can be merged.

@fisker Which do you prefer or think is easier to be maintained?

I've tested it somewhat in the playground and this code appears to me to have no logical errors or overlooked considerations.

fisker commented:
> Two if statements can be merged.

👍

byplayer committed:
Update src/language-markdown/print/children.js

Co-authored-by: Tatsunori Uchino <tats.u@live.jp>

byplayer committed:
Update src/language-markdown/print/children.js

Co-authored-by: Tatsunori Uchino <tats.u@live.jp>

byplayer commented:
I applied two suggestions.
Thank you.

autofix-ci[bot] committed:
Apply automated changes

tats-u commented on src/language-markdown/print/children.js:
```diff
@@ -71,7 +71,10 @@ function shouldPrePrintDoubleHardline(path, options) {
       (node.type === "list" &&
         parent.type === "listItem" &&
-        previous.type === "code")
+        (previous.type === "code" ||
+          // Preserve blank line before nested list within listItem (issue #17746)
+          previous.type === "paragraph") &&
+        previous.position.end.line + 1 < node.position.start.line)
```
Do you have an idea to test the path `previous.type === "code" && previous.position.end.line + 1 < node.position.start.line`? No newly added testcases contain `"code"` (code blocks).

tats-u replied:
I found one:

````md
- a
  ```
  b
  ```
- c

  - d
````

https://spec.commonmark.org/dingus/?text=-%20a%0A%20%20%60%60%60%0A%20%20b%0A%20%20%60%60%60%0A-%20c%0A%0A%20%20-%20d%0A%0A---%0A%0A-%20a%0A%20%20%60%60%60%0A%20%20b%0A%20%20%60%60%60%0A-%20c%0A%20%20-%20d

https://deploy-preview-17857--prettier.netlify.app/playground/?version=next#N4Igxg9gdgLgprEAuEBaABAQwDpXegAyN3wCMTDioMxcKMATEAGhAgAcYBLaAZ2VCYATkIgB3AArCE-FJgA2YzAE9+rUkMxgA1nBgBlTAFs4AGS5Q4yAGYLecVhFIArOGBgB1Te2Qh2QuHshADcrdU0dPX12LQsAcxs7BxB7Iy5kGCEAV2S4AA92OCEuE1gFAHlCzRgIIQkIXi5uaF8EJlZ8qpKEGAUAFSKoYS5AxPl7VkaoOPk4AEUsiHgxiZBnXjz9eNmFpaskW3HkgEdF+AlRdlkQTF5USzgGR5YQTMwueXiAYQgjI0xfAp5C8pjM4ABBGCZLikLLnIrmSwrZIACxgRnkHhRTUCMTAcH0MiaXGCTWUvjAvDUIGCOQAklAnrB9GBipxwYz9DBlLNkax-A04F5MD4UP5AkVQi82uVrIj9iB5NYXhYgjALpg4v8+X5hEFfP8hNoGOIoC9-BZPFwGDAUcgABwABlYAVOXACGq1AIOSVYvVIHmttuQACZWFl7H1MKRZIdVnAjKRHk8GKZMNMspq4AAxWr-KHxQFwiAgAC+paAA

byplayer replied:
Thank you for your comment.
I added test case.

byplayer committed:
✅ test(markdown): cover code→nested-list blank-line path for #17746

# Why is this change needed?
The `previous.type === "code" && previous.position.end.line + 1 <
node.position.start.line` branch in `shouldPrePrintDoubleHardline`
(both the markdown and MDX paths) had no dedicated fixture. Prior
snapshots only exercised the tight (no-blank) negative path via
`issue-17746-code-before-list`, leaving the positive "preserve the
blank line" path untested for code blocks. A PR reviewer on #17857
pointed this out and suggested a repro.

# How does it address the issue?
Added fixtures that place a code block followed by a blank line and
a nested list inside the same listItem, plus the reviewer's mixed
case (sibling listItem with code + paragraph/nested-list with blank
line). Both fenced and indented code variants are covered, in both
markdown and MDX parsers.

- tests/format/markdown/list/issue-17746-code-sibling-nested-list.md
- tests/format/markdown/list/issue-17746-fenced-code-then-nested-list.md
- tests/format/markdown/list/issue-17746-indented-code-then-nested-list.md
- tests/format/mdx/mdx/issue-17746-fenced-code-then-nested-list.mdx
- tests/format/mdx/mdx/issue-17746-indented-code-then-nested-list.mdx

Refs: #17746, #17857

tats-u commented on tests/format/markdown/list/issue-17746-fenced-code-then-nested-list.md:
````diff
@@ -0,0 +1,5 @@
+  a
+  ```
+
+  - b
````
Nitpick: as it stands now, the HTML conversion result wouldn't change even when the blank line above `- b` were removed. How about:

```suggestion
  - b
- c
```

tats-u reviewed (approved):
LGT*M*

@fisker

fisker committed:
Apply suggestions from code review

Co-authored-by: Tatsunori Uchino <tats.u@live.jp>

fisker committed:
Update snapshots
